### DISCUSSION TITLE
SysVinit Script that ships with Debian Bullseye '.deb' of Radicale3 does not work.

### DISCUSSION THREAD

**Discussion by jnr2820** [Kozea/Radicale#1414] [External Packaging]
Hi All,

The SysVinit script that ships with the Debian Bullseye '.deb' of Radicale3 does not work for systems running SysV instead of SystemD. (Service fails to start). It appears to be the same as the the one that comes with the '.deb' for Radicale2 (Buster). It calls the "--daemon" flag which is valid for Radicale2 but doesn't appear to be valid for Radicale3. Simply removing "--daemon" option does not help.

Has support for SysVinit been totally dropped? Is this now something that the package maintainer should manage?
Are there any plans to rewrite the script for Radicale3? (It would be greatly appreciated. :) )

More info: 
Devuan 3 "Beowolf" (Debian Buster with SysVinit) - Radicale2 starts fine.
Devuan 4 "Chimaera" (Debian Bullseye with SysVinit) - Radicale3 fails to start. ("/etc/init.d/radicale" appears to be the same as the one from Radicale2, but seems to need to a rewrite to work with Radicale3)

```
jeremiah@devuan-vbox-lxde:~$ sudo service radicale start
Starting Radicale CalDAV server : radicale.
jeremiah@devuan-vbox-lxde:~$ sudo service radicale status
radicale is not running ... failed!
```
```
jeremiah@devuan-vbox-lxde:~$ sudo -i -u radicale 
radicale@devuan-vbox-lxde:~$ radicale -D
[2021-10-19 17:08:22 -0400] [3124] [INFO] Loaded default config
[2021-10-19 17:08:22 -0400] [3124] [INFO] Loaded config file '/etc/radicale/config'
[2021-10-19 17:08:22 -0400] [3124] [INFO] Skipped missing config file '/var/lib/radicale/.config/radicale/config'
[2021-10-19 17:08:22 -0400] [3124] [INFO] Loaded arguments
[2021-10-19 17:08:22 -0400] [3124] [INFO] Starting Radicale
[2021-10-19 17:08:22 -0400] [3124] [INFO] auth type is 'radicale.auth.remote_user'
[2021-10-19 17:08:22 -0400] [3124] [INFO] storage type is 'radicale.storage.multifilesystem'
[2021-10-19 17:08:22 -0400] [3124] [INFO] rights type is 'radicale.rights.from_file'
[2021-10-19 17:08:22 -0400] [3124] [INFO] web type is 'radicale.web.internal'
[2021-10-19 17:08:22 -0400] [3124] [INFO] Listening on '[127.0.0.1]:5232'
[2021-10-19 17:08:22 -0400] [3124] [INFO] Listening on '[::1]:5232'
[2021-10-19 17:08:22 -0400] [3124] [INFO] Radicale server ready
```
As you can see, it starts up fine ran as the 'radicale' user that script is configured to use.
Init script attached as TXT file: [radicale_init.txt](https://github.com/Kozea/Radicale/files/7376686/radicale_init.txt)

**Comment by jnr2820** [comment]
I noticed after submitting this that the Init Script was written by a Package Maintainer. I filed a bug via "reportbug" within Devuan. Hopefully they see it and forward it upstream if needed.

That may take some time, so I'm open to ideas if y'all have any adapted Init Scripts that y'all have been using with Radicale 3. I'll post the upstream bug number if/when I get it.

**Comment by ehem** [comment]
Init scripts shipped with Debian are supposed to work, so this seemed better reported to the Debian Project.  Bug [#1000147](https://bugs.debian.org/1000147) has been marked as fixed (as someone was able to figure out working adjustments).

Another issue is the complete removal of logging support.  System daemons aren't supposed to be logging via stdout/stderr, so Radicale 3.0 dumping all messages to output is a problem.  Perhaps SystemD supports this, but last I'd read this software was supposed to be "Radicale", not "SystemD-Radicale".

**Comment by jonassmedegaard** [comment]
Quoting Elliott Mitchell (2021-12-06 02:52:57)
> Another issue is [...]

Please report each issue as a separate report, to make them actionable.

Or mention issue ID for an issue already reported.

**Comment by jnr2820** [comment]
Hey @ehem ,
Does the latest built in backports ("3.1.7-1~bpo11+1") work for you? I did a clean install in a VM of Devuan 4 and I am still not able to get the service to start. Did it not get fixed in Deb Bug [#1000147](https://bugs.debian.org/1000147) ?

**Comment by Jorilx** [comment]
I fixed the calls to `start-stop-daemon` by replacing the `--name $NAME` with `--exec $DAEMON` and now I can get Radicale v3 to start and stop correctly. Here's the patch:
```
--- radicale.ori	2022-08-11 21:12:44.855323769 +0200
+++ radicale	2022-08-16 09:00:08.335463039 +0200
@@ -68,10 +68,10 @@
 	fi
 
 	start-stop-daemon --start --quiet --startas $DAEMON \
-		--name $NAME --test > /dev/null \
+		--exec $DAEMON --test > /dev/null \
 		|| return 1
 	start-stop-daemon --start --quiet --startas $DAEMON --background \
-		--name $NAME --umask 0027 --chuid $DAEMON_UID:$DAEMON_GID -- \
+		--exec $DAEMON --umask 0027 --chuid $DAEMON_UID:$DAEMON_GID -- \
 		$RADICALE_OPTS \
 		|| return 2
 }
```

There's still the problem with logging, though

**Comment by Jorilx** [comment]
Since we're already using the `--background` switch we can add `--no-close` and redirect stderr to a log file... So here's the "final" patch:
```
--- radicale.ori	2022-08-11 21:12:44.855323769 +0200
+++ radicale	2022-08-16 09:53:49.885187840 +0200
@@ -68,11 +68,11 @@
 	fi
 
 	start-stop-daemon --start --quiet --startas $DAEMON \
-		--name $NAME --test > /dev/null \
+		--exec $DAEMON --test > /dev/null \
 		|| return 1
-	start-stop-daemon --start --quiet --startas $DAEMON --background \
-		--name $NAME --umask 0027 --chuid $DAEMON_UID:$DAEMON_GID -- \
-		$RADICALE_OPTS \
+	start-stop-daemon --start --quiet --startas $DAEMON --background --no-close \
+		--exec $DAEMON --umask 0027 --chuid $DAEMON_UID:$DAEMON_GID -- \
+		$RADICALE_OPTS >> /var/log/radicale.log 2>&1 \
 		|| return 2
 }

```

**Comment by Jorilx** [comment]
It looks like the stop/restart functions need a little fix too:
```
--- radicale.ori	2022-08-11 21:12:44.855323769 +0200
+++ radicale	2022-08-16 11:09:04.446920188 +0200
@@ -86,7 +86,7 @@
 	#   1 if daemon was already stopped
 	#   2 if daemon could not be stopped
 	#   other if a failure occurred
-	start-stop-daemon --stop --quiet --retry=TERM/30/KILL/5 --name $NAME
+	start-stop-daemon --stop --quiet --retry=TERM/30/KILL/5 --user $DAEMON_UID
 }
 
 case "$1" in
```

**Comment by kosh-mclien** [comment]
As I understand it the bug-report for this one is addressed here:
[#1038930](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1038930) (existing patch, not implemented)
But there is still a patch missing for Version: 3.1.8-2:
[#1055771](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1055771) (no patch so far)

**Comment by bughunter2** [comment]
Note that the dpkg start-stop-daemon supports the --output option to redirect output streams (both stdout and stderr).

I came up with the following patch and looked where to submit it and found this thread. Hence I'll just leave this here.

```
--- radicale.original	2022-11-16 20:15:19.000000000 +0000
+++ radicale.proposedfix	2024-06-12 20:41:57.389487566 +0000
@@ -22,6 +22,7 @@
 RUNDIR=/run/$NAME
 CALDIR=/var/lib/$NAME
 LOGDIR=/var/log/$NAME
+LOGFILE=/var/log/$NAME/$NAME.log
 SCRIPTNAME=/etc/init.d/$NAME
 
 # Exit if the package is not installed
@@ -70,8 +71,9 @@
 	start-stop-daemon --start --quiet --startas $DAEMON \
 		--exec $DAEMON --test > /dev/null \
 		|| return 1
-	start-stop-daemon --start --quiet --startas $DAEMON --background \
-		--exec $DAEMON --umask 0027 --chuid $DAEMON_UID:$DAEMON_GID -- \
+	start-stop-daemon --start --quiet --startas $DAEMON \
+		--exec $DAEMON --umask 0027 --chuid $DAEMON_UID:$DAEMON_GID \
+		--background --no-close --output $LOGFILE -- \
 		$RADICALE_OPTS \
 		|| return 2
 }

```

**Reply by jonassmedegaard** [reply]
Thank you, @bughunter2.

Please in future, however - also everyone else reading this - report bugs in redistribution code at that distribution. It might seem unimportant, but tracking issues where they belong really helps ensuring the highest quality maintenance with the least effort.

For Debian, specifically, [there is a web page documenting how to report issues](https://www.debian.org/Bugs/Reporting).

**Reply by bughunter2** [reply]
Ah, I must've assumed this was the place, given the discussion. No problem. I have replied at [debian bug report 1038930](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1038930). Cheers! :slightly_smiling_face:
